### PR TITLE
Open amp refactoring and add more platforms support

### PR DESCRIPTION
**Pull request by karl-zh**
openAMP is a multiple core IPC which supported in Zephyr. This PR is going to make it supported for more multiple core platforms rather than for lpc54 only.

Currently added the arm platform AN521 and Musca A, for further plan is to supporting Musca B1 when #16729 get merged.

### CODE REVIEW

**Comment by galak** [comment]
@arnop2 can you take a look at this PR.  I'm not that familiar with how we should be conveying the shared memory region in DTS.

File: boards/arm/mps2_an521/mps2_an521_nonsecure.dts
Comment (arnop2) [inline boards/arm/mps2_an521/mps2_an521_nonsecure.dts]: Very interesting implementation based on the memory declaration in the DTS. I hesitated to do same in my pull request (https://github.com/zephyrproject-rtos/zephyr/pull/16985).
A drawback in this implementation is that the memory region is project dependent. It is needed only for OpenAMP, but declared for a board, so for all projects.
For instance, this result here in an overlap between the memory declared for the zephyr firmware and this shared memory. not safe...
An other alternative would be  to limit the zephyr SRAM, but impact is for all projects...

So a point to solve here (IMHO) is to make this shared region optional ( project dependent) to allow to use it for some other purposes. But also to configure the rest of the ram according to this shared memory.

To get around the issue, In my pull request, i have implemented the memory pool  as a linker memory region, under CONFIG_OPENAMP. But I'm not satisfied with my implementation. If my shared memory is optional, This does not prevent overlap with the zephyr SRAM. 

I have no real solution  to propose:), i can just try to clarify the equation:
- There is a dependency on project: use of a CONFIG_OPENAMP?
- There is a dependency on the platform ( memory region definition + Zephyr RAM resizing): DTS overlay, linker with some CONFIG values, ...?

File: boards/arm/mps2_an521/mps2_an521.dts
Comment (arnop2) [inline boards/arm/mps2_an521/mps2_an521.dts]: strange address regarding other dts files of you pull request...Is it a specific normal that this memory is  not include in sram0?

File: samples/subsys/ipc/openamp/README.rst
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.rst]: should ``slave`` be ``remote``, since that's what shown in the sample output below?

**Review by dbkinder — CHANGES_REQUESTED** [review]
one doc question, otherwise looks good.

File: boards/arm/mps2_an521/mps2_an521.dts
Comment (karl-zh) [reply boards/arm/mps2_an521/mps2_an521.dts]: Yes, this seems strange, not included in the normal sram and also for all projects rather than openAMP itself.

Then, how about define the shared memory address as a macro in soc.h with the control of CONFIG_OPENAMP defined?

If using CONFIG_SHM_ADDR_xxx, that may require more prj.conf files for platfrom specific I think.

@galak

Comment (arnop2) [reply boards/arm/mps2_an521/mps2_an521.dts]: I don't know, and probably I'm not expert enough in zephyr to determine the best solution.....
For a project integrating OpenAMP, I assume that the good practice is to define the shared memory in DTS and to resize the zephyr RAM accordingly. 
I don't know if it would be possible to create a dts overlay associated with the example project...? I couldn't find any examples.

Comment (arnop2) [reply boards/arm/mps2_an521/mps2_an521.dts]: Sorry I'm Linux formatted, so i looked for .dtso files... overlay is possible for sample ( but maybe you already know this). There are a lot of .overlay files in samples. so you could use similar implementation. 
For instance ( code not tested) you could create a  mps2_an521_nonsecure.overlay:
&sram0: {
		reg = <0x28100000 0xF80000>;
	};

/ {
	sramx: memory@0x281F8000 {
		compatible = "mmio-sram";
		reg = <0x281F8000 0x8000>;
	};
};

This limits the RAM used by the linker script, so ensure coherency in memory allocations, and allow to declare the memory assigned to the OpenAMP  in an elegant way. So it seems like a good compromise (IMHO).

File: samples/subsys/ipc/openamp/README.rst
Comment (karl-zh) [reply samples/subsys/ipc/openamp/README.rst]: Hi David
Fixed now. Thanks for your review.

File: boards/arm/mps2_an521/mps2_an521.dts
Comment (karl-zh) [reply boards/arm/mps2_an521/mps2_an521.dts]: Hi Arnaud,

Thanks for your suggestion of overlay. I updated it now.

File: soc/arm/arm/mps2/dts_fixup.h
Comment (arnop2) [inline soc/arm/arm/mps2/dts_fixup.h]: With overlays in sample you need to use a chosen{} defined in the overlay instead
(https://docs.zephyrproject.org/latest/guides/dts/index.html?highlight=chosen#aliases-and-chosen-nodes)

**Review by arnopo — CHANGES_REQUESTED** [review]

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

File: samples/subsys/ipc/openamp/sse_200.conf
Comment (karl-zh) [reply samples/subsys/ipc/openamp/sse_200.conf]: Okay, the IPM MHU could be in soc or board level I think.
It seems still have some specific for lpcx, like:
`CONFIG_IPM_MCUX=y
CONFIG_SLAVE_CORE_MCUX=y
CONFIG_SLAVE_IMAGE_MCUX="${ZEPHYR_BINARY_DIR}/../openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin"`

sse_200.conf is for AN521, Musca A and Musca B for common.

Comment (arnop2) [reply samples/subsys/ipc/openamp/sse_200.conf]: The stm32H7 Soc should be able to reuse it also (need TBC).
The idea here would be  to have something similar to https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/bluetooth/hci_uart . A generic conf file that can be used by default or used as template to adapt to a specific platform need.
This is of course  a proposal and not a request...

File: soc/arm/arm/mps2/dts_fixup.h
Comment (arnop2) [reply soc/arm/arm/mps2/dts_fixup.h]: @karl-zh  
I updated my PR (https://github.com/zephyrproject-rtos/zephyr/pull/16985). I implemented the DT overlay in the sample with the use of chosen{} to have a generic name for the shared memory pool.  To do similar implementation, you have to get this commit : https://github.com/zephyrproject-rtos/zephyr/pull/16985/commits/19309c7a5608b6c2b6f06dab9fceb9edba8a9acf. Then you have to create the overlay using the template <board>.overlay( including the memory and the chosen declarations ). The overlay will be automatically applied on top of and DT_IPC_SHM_BASE_ADDRESS + DT_IPC_SHM_BASE_SIZE defines will be generated.

@galak could you confirm if you are in line with this DT overlay approach

**Review by arnopo — CHANGES_REQUESTED** [review]
Just 2 minor remarks remaining, then  should be ok for me

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (karl-zh) [reply samples/subsys/ipc/openamp/CMakeLists.txt]: Fine, I was thinking to put all overlay files into a folder, since there will be more platforms supported. Will update later.

File: samples/subsys/ipc/openamp/common.h
Comment (karl-zh) [reply samples/subsys/ipc/openamp/common.h]: Hi Arnaud:
Okay, this is derived from the original demo, it is a shared memory based on ram. So I think it is make sense to continue use "sramx.shm". Do you have some ideas on this?
Thanks

Comment (arnop2) [reply samples/subsys/ipc/openamp/common.h]: Just the "sramx" which can make no sense depending on platform. 
 " sh_mem" could be more generic, but feel free to choice a better name :)

File: samples/subsys/ipc/openamp/CMakeLists.txt
Comment (arnop2) [reply samples/subsys/ipc/openamp/CMakeLists.txt]: Reading my comment, i not sure that my comment is clear... so in doubt:
When i say the "overlay file is automatically set", i mean DTC_OVERLAY_FILE environment variable is automatically set.

Comment (karl-zh) [reply samples/subsys/ipc/openamp/CMakeLists.txt]: Yes. Thanks, I have changed this to using automatically.

File: scripts/dts/extract/globals.py
Comment (erwango) [inline scripts/dts/extract/globals.py]: Expect conflicts with #17660

Comment (erwango) [reply scripts/dts/extract/globals.py]: #17660 is now merged, you'll have to report this patch in `scripts/dts/gen_defines.py`

Comment (arnop2) [reply scripts/dts/extract/globals.py]: related commit updated in https://github.com/zephyrproject-rtos/zephyr/pull/16985

Comment (karl-zh) [reply scripts/dts/extract/globals.py]: Updated.

**Review by arnopo — APPROVED** [review]
Just a minor request to change SHM_DEVICE_NAME definition + a rebase to fix my patch commit messag ( apology for this).
Then LGTM, thanks.

**Review by erwango — APPROVED** [review]

**Comment by karl-zh** [comment]
> Just a minor request to change SHM_DEVICE_NAME definition + a rebase to fix my patch commit messag ( apology for this).
> Then LGTM, thanks.

Thanks. Let us open for more ideas, try to make everyone happy on this. 👍

File: samples/subsys/ipc/openamp/remote/src/main.c
Comment (galak) [inline samples/subsys/ipc/openamp/remote/src/main.c]: Why does the data here mater?

Comment (karl-zh) [reply samples/subsys/ipc/openamp/remote/src/main.c]: CPU0 send notification to CPU1, vise vesa.

Comment (karl-zh) [reply samples/subsys/ipc/openamp/remote/src/main.c]: The dummy data is not necessary here for MHU based hardware. But it needs to distinguish the current core No. for notifications.

**Comment by galak** [comment]
Ok, I tried this on musca-a1 and qemu-mps2_an521 and it doesn't seem to work.  Also tried the `samples/subsys/ipc/ipm_mhu_dual_core` on those platforms and they didn't seem to pass.  Something to debug this week.

**Comment by karl-zh** [comment]
> Ok, I tried this on musca-a1 and qemu-mps2_an521 and it doesn't seem to work. Also tried the `samples/subsys/ipc/ipm_mhu_dual_core` on those platforms and they didn't seem to pass. Something to debug this week.

Okay, I also need some refresh of memory to this context. BTW, do you have some fail context? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:

**AN521:** 
~/zephyr/samples/subsys/ipc/openamp/build$ rm -rf *; rm -rf .* ; cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf  ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin

Config file:
TITLE: Versatile Express Images Configuration File

[IMAGES]
TOTALIMAGES: 2                     ;Number of Images (Max: 32)

IMAGE0ADDRESS: 0x10000000          ;Please select the required executable program
IMAGE0FILE: \SW\zephyr.bin  ; - selftest

IMAGE1ADDRESS: 0x100000
IMAGE1FILE: \SW\zephyr_ns.bin

**Comment by galak** [comment]
> Okay, I also need some refresh of memory to this context. Maybe some rebase conflicts or other changes need rebase. BTW, do you have some fail details? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:
> 
> **AN521:**
> ~/zephyr/samples/subsys/ipc/openamp/build$ cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin
> 
> Config file:
> TITLE: Versatile Express Images Configuration File
> 
> [IMAGES]
> TOTALIMAGES: 2 ;Number of Images (Max: 32)
> 
> IMAGE0ADDRESS: 0x10000000 ;Please select the required executable program
> IMAGE0FILE: \SW\zephyr.bin ; - selftest
> 
> IMAGE1ADDRESS: 0x100000
> IMAGE1FILE: \SW\zephyr_ns.bin

Hmm, need to parse this in more detail.  However is it possible to run the test w/o SECURE/NON-SECURE?

**Comment by karl-zh** [comment]
> > Okay, I also need some refresh of memory to this context. Maybe some rebase conflicts or other changes need rebase. BTW, do you have some fail details? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:
> > **AN521:**
> > ~/zephyr/samples/subsys/ipc/openamp/build$ cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin
> > Config file:
> > TITLE: Versatile Express Images Configuration File
> > [IMAGES]
> > TOTALIMAGES: 2 ;Number of Images (Max: 32)
> > IMAGE0ADDRESS: 0x10000000 ;Please select the required executable program
> > IMAGE0FILE: \SW\zephyr.bin ; - selftest
> > IMAGE1ADDRESS: 0x100000
> > IMAGE1FILE: \SW\zephyr_ns.bin
> 
> Hmm, need to parse this in more detail. However is it possible to run the test w/o SECURE/NON-SECURE?

I might using a wrong binary name here. They are following the start work for tfm integrating with Zpehyr.
The two binaries here are for different CPUs, \SW\zephyr.bin is for CPU0 and \SW\zephyr_ns.bin is for CPU1,  the name is for configurations file only(images.txt).

**Comment by jhedberg** [comment]
This PR seems to require rebasing since there are conflicts. Is something still missing before it can be merged (e.g. approvals or some additional code changes)?

**Comment by karl-zh** [comment]
> This PR seems to require rebasing since there are conflicts. Is something still missing before it can be merged (e.g. approvals or some additional code changes)?

Hi Johan

I think this PR needs approval before the merge.

**Comment by jhedberg** [comment]
> I think this PR needs approval before the merge.

@karl-zh while waiting for approval, could you rebase and fix the conflict with `samples/subsys/ipc/openamp/sample.yaml`?

**Comment by karl-zh** [comment]
> > I think this PR needs approval before the merge.
> 
> @karl-zh while waiting for approval, could you rebase and fix the conflict with `samples/subsys/ipc/openamp/sample.yaml`?

Hi
 I was trying to resolve the conflict from github, but it seems CI fail. I am on leave for the Chinese New Year holiday recently and will rebase/resolve and try to make CI pass when back to the office. 
The approval and review is required too.

Thanks for your comments.

**Comment by erwango** [comment]
@karl-zh, was this intentional to push a merge commit ?

**Review by galak — APPROVED** [review]

**Pull request merged**
